### PR TITLE
Add generic associatedtypes and generalized supertype constraints

### PR DESCRIPTION
**Pull request by anandabits**
This PR adds a few items to the generics manifesto.  The first two were discussed briefly on Twitter where @DougGregor requested a PR to add them.  I plan to followup with a pitch to Swift evolution for generalized supertype constraints after this is merged.

### CODE REVIEW

File: docs/GenericsManifesto.md [hunk 1]
```diff
@@ -87,6 +87,21 @@ typealias StringDictionary<Value> = Dictionary<String, Value>
 var d1 = StringDictionary<Int>()
 var d2: Dictionary<String, Int> = d1 // okay: d1 and d2 have the same type, Dictionary<String, Int>
 ```
+### Generic associatedtypes
```
Comment (slavapestov) [inline docs/GenericsManifesto.md]: "associated types" here and below

Comment (anandabits) [reply docs/GenericsManifesto.md]: That’s how I would usually write it but I followed the precedent of “typealiases” that’s already in this document.

Comment (slavapestov) [reply docs/GenericsManifesto.md]: I think that should be fixed too. In diagnostics we tend to say "type alias" not "typealias"

File: docs/GenericsManifesto.md [hunk 2]
```diff
@@ -592,6 +622,43 @@ one cannot directly specify `T`: either one calls `f` (and `T` is determined via
 ```Swift
 let x = f<Int> // x has type (Int) -> Void
 ```
+### Conditional default arguments
```
Comment (xwu) [inline docs/GenericsManifesto.md]: Drive-by nit: please add an empty line above this.

Comment (slavapestov) [inline docs/GenericsManifesto.md]: Has this feature been discussed previously anywhere? I don't think we should add it to the document. The other two are "canonical" generics features that we have all talked about before.

Comment (anandabits) [reply docs/GenericsManifesto.md]: Not to my knowledge.  It was on my mind while I was writing this so I added it.  I can remove it if others agree that it shouldn’t be added.

File: docs/GenericsManifesto.md (hunk 1 above)
Comment (anandabits) [reply docs/GenericsManifesto.md]: Do you suggest I include fixes for all of them in this PR?

File: docs/GenericsManifesto.md [hunk 3]
```diff
@@ -249,6 +264,21 @@ typealias AnyObject = protocol<class>
...
+Currently, supertype constraints may only be specified using a concrete class or protocol type.  This prevents us from abstracting over the supertype.
+
+```Swift
+protocol P {
+  associatedtype Base
+  associatedtype Derived: Base
+}
+```
+
+In the above example `Base` may be any non-final class or existential type.  If `Base` is a class `Derived` must inherit from `Base`.  If `Base` is an existential `Derived` may be any type that provides a compatible conformance, including a value type.
```
Comment (DougGregor) [inline docs/GenericsManifesto.md]: I had assumed that we were talking about a superclass constraint here, but I *think* it generalizes to existentials correctly. It's slightly bothersome to have the requirement on `Base` be "non-final class or existential type", and I wonder if perhaps there's a looser definition that's okay. For example, maybe `Base == Derived` is always okay, even if `Base` is (say) a value type.

File: docs/GenericsManifesto.md (hunk 2 above)
Comment (DougGregor) [reply docs/GenericsManifesto.md]: I guess I'd rather have it removed from this PR and discuss it on swift-evolution a little bit first to get some basic syntax ideas in place. And I do want to discuss this... it's something that we might be able to use to help simplify associated type inference.

File: docs/GenericsManifesto.md (hunk 3 above)
Comment (anandabits) [reply docs/GenericsManifesto.md]: That's good feedback.  I didn't word this precisely enough.  I think it makes sense to allow `Base` to be any type.  `Derived` may be the same as `Base`, a subclass of `Base` when `Base` is a non-final class or a conforming type when `Base` is an existential.  Does that loosening make sense?

File: docs/GenericsManifesto.md (hunk 2 above)
Comment (anandabits) [reply docs/GenericsManifesto.md]: Sounds good.  I'll remove it when I update the PR and send a separate pitch on evolution.

I was thinking of this primarily as a way to reduce the pain of providing an interface like this but if it can help simplify associated type inference it would move up the priority list for me quite a bit.  I have been seeing associated type inference issues in code that uses an overload set like this.

File: docs/GenericsManifesto.md (hunk 1 above)
Comment (anandabits) [reply docs/GenericsManifesto.md]: @DougGregor any thoughts on this?  Should I make the change from "associatedtypes"  to "associated types" and "typealiases" to "type aliases" in this PR?

File: docs/GenericsManifesto.md (hunk 3 above)
Comment (DougGregor) [reply docs/GenericsManifesto.md]: Yes, that makes sense to me.

Comment (karwa) [reply docs/GenericsManifesto.md]: Would it be clearer to refer to these as "is" constraints? Derived may be any type so long as values of that type return true from a `is Base` condition.

File: docs/GenericsManifesto.md (hunk 1 above)
Comment (karwa) [reply docs/GenericsManifesto.md]: It's probably worth a small note about how this might look from the perspective of somebody trying to conform to a protocol with generic associated types. The way I guess it would work is that a generic associated-type would require a generic type alias in the conforming type. So the generic type which fulfils that requirement may have more generic parameters, as long as they are all bound to parameters in the parent type or to concrete types.

So in your example, `Wrapped<T>` could be a tuple of `(Int, T)`, or an `Array<T>`, or a `Dictionary<SomeNestedGenericParameter, T>`. Is that correct?

Comment (anandabits) [reply docs/GenericsManifesto.md]: @karwa This is a good idea.  I'll be updating the PR with a note about this shortly.  Your understanding is correct.

**Comment by anandabits** [comment]
@DougGregor I have removed the discussion of conditional default arguments and updated the discussion of generalized supertype constraints.

**Comment by DevAndArtist** [comment]
This really should be merged 👍

**Comment by AndrewSB** [comment]
yearly reminder to merge

**Comment by CodaFi** [comment]
@DougGregor Is this still a direction we want to go in?

**Comment by DevAndArtist** [comment]
I personally would wish that at least generalized super type constraint would eventually become a thing in Swift. I bumped too many times where I couldn‘t tell the compiler some natural context related sub-type relation, which requires a runtime workaround instead of a compile-time type-safety.

**Comment by CodaFi** [comment]
Got the green light from Doug

⛵️ 

@swift-ci please smoke test and merge
